crates/wasmtime/src/runtime/ref/gc_ref.rs

```diff
@@ -1,52 +1,318 @@
-use crate::AsContextMut;
+mod rooting;
+
+use anyhow::anyhow;
+pub use rooting::*;
+
+use crate::{store::StoreOpaque, AsContext, AsContextMut, Result};
 use std::any::Any;
 use std::ffi::c_void;
-use wasmtime_runtime::VMExternRef;
+use wasmtime_runtime::{VMExternRef, VMGcRef};
+
+/// Upcast a `VMExternRef` into a `VMGcRef`.
+pub(crate) fn extern_ref_into_gc_ref(e: VMExternRef) -> VMGcRef {
+    unsafe { VMGcRef::from_ptr(e.into_raw()).unwrap() }
+}
+
+/// Downcast a `VMGcRef` into a `VMExternRef`.
+///
+/// # Safety
+///
+/// The given `gc_ref` must be a valid `VMExternRef`.
+///
+/// The resulting `VMExternRef` has non-atomic reference counting, and must only
+/// be used within the context of a mutable store/GC heap.
+pub(crate) unsafe fn extern_ref_from_gc_ref(gc_ref: VMGcRef) -> VMExternRef {
+    VMExternRef::from_raw(gc_ref.as_non_null().as_ptr()).unwrap()
+}
 
-/// Represents an opaque reference to any data within WebAssembly.
-#[derive(Clone, Debug)]
+/// Clone a `VMExternRef` from a `VMGcRef`.
+///
+/// This requires mutable access to the store to ensure that the underlying
+/// non-atomic reference counting is safe.
+///
+/// # Safety
+///
+/// The given `gc_ref` must be a valid `VMExternRef`.
+///
+/// The `gc_ref` must be associated with this store.
+pub(crate) unsafe fn clone_extern_ref_from_gc_ref(
+    _store: &mut StoreOpaque,
+    gc_ref: VMGcRef,
+) -> VMExternRef {
+    VMExternRef::clone_from_raw(gc_ref.as_non_null().as_ptr()).unwrap()
+}
+
+unsafe fn extend_to_store_lifetime<'a, 'b, T>(_store: &'a impl AsContext, reference: &'b T) -> &'a T
+where
+    'a: 'b,
+    T: ?Sized,
+{
+    std::mem::transmute(reference)
+}
+
+unsafe fn extend_to_store_lifetime_mut<'a, 'b, T>(
+    _store: &'a mut impl AsContextMut,
+    reference: &'b mut T,
+) -> &'a mut T
+where
+    'a: 'b,
+    T: ?Sized,
+{
+    std::mem::transmute(reference)
+}
```

Comment:
Personally I'm very wary of these functions in addition to the `ref_from_raw` functions in the runtime crate. Can these be removed entirely? I would like to ideally rely on natural lifetime stuff in rustc to avoid any need to do this sort of transmute.

For example there's basically no way to locally reason that this `transmute` is correct, so I need to build a total understanding of the whole GC as a result to know when it's safe and when it's not safe to use this.